networkx/algorithms/tests/test_swap.py

```diff
@@ -12,13 +12,31 @@ def test_directed_edge_swap():
     assert out_degrees == sorted((n, d) for n, d in G.out_degree())
 
 
+def test_edge_cases_directed_edge_swap():
+    e = (
+        f"Maximum number of swap attempts \\(11\\) exceeded "
+        f"before desired swaps achieved \\(4\\)."
+    )
+    graph = nx.path_graph(4, create_using=nx.DiGraph)
+    with pytest.raises(nx.NetworkXAlgorithmError, match=e):
+        nx.directed_edge_swap(graph, nswap=4, max_tries=10, seed=1)
+    graph = nx.DiGraph([(0, 0), (0, 1), (1, 0), (2, 3), (3, 2)])
+    with pytest.raises(nx.NetworkXAlgorithmError, match=e):
+        nx.directed_edge_swap(graph, nswap=4, max_tries=10, seed=1)
```

Comment:
IIUC, the exception raised here is actually a bit misleading as the edge swapping is impossible for this case. We can still add the test for the sake of coverage, but let's also add a comment explaining it a bit. In principle, someone could (in a perfect world) adjust the `directed_edge_swap` function to detect when an edge swap isn't possible, and report a better error. If/when this happens, we'd want to be able to identify this test case and know that it's safe to delete!